ast/compile.go

```diff
@@ -4243,60 +4243,68 @@ func appendExpr(body Body, expr *Expr) Body {
+	result, terms[2] = rewriteDynamicsInTerm(expr, f, bi, terms[2], result)
 	return appendExpr(result, expr)
 }
 
-func rewriteDynamicsCallExpr(f *equalityFactory, expr *Expr, result Body) Body {
+func rewriteDynamicsCallExpr(f *equalityFactory, bi map[string]*Builtin, expr *Expr, result Body) Body {
 	terms := expr.Terms.([]*Term)
 	for i := 1; i < len(terms); i++ {
-		result, terms[i] = rewriteDynamicsOne(expr, f, terms[i], result)
+		_, isBuiltin := refBuiltins[terms[0].String()]
+		result, terms[i] = rewriteDynamicsOneKeepRefs(expr, f, bi, terms[i], result, isBuiltin)
 	}
 	return appendExpr(result, expr)
 }
 
-func rewriteDynamicsEveryExpr(f *equalityFactory, expr *Expr, result Body) Body {
+func rewriteDynamicsEveryExpr(f *equalityFactory, bi map[string]*Builtin, expr *Expr, result Body) Body {
 	ev := expr.Terms.(*Every)
-	result, ev.Domain = rewriteDynamicsOne(expr, f, ev.Domain, result)
-	ev.Body = rewriteDynamics(f, ev.Body)
+	result, ev.Domain = rewriteDynamicsOne(expr, f, bi, ev.Domain, result)
+	ev.Body = rewriteDynamics(f, bi, ev.Body)
 	return appendExpr(result, expr)
 }
 
-func rewriteDynamicsTermExpr(f *equalityFactory, expr *Expr, result Body) Body {
+func rewriteDynamicsTermExpr(f *equalityFactory, bi map[string]*Builtin, expr *Expr, result Body) Body {
 	term := expr.Terms.(*Term)
-	result, expr.Terms = rewriteDynamicsInTerm(expr, f, term, result)
+	result, expr.Terms = rewriteDynamicsInTerm(expr, f, bi, term, result)
 	return appendExpr(result, expr)
 }
 
-func rewriteDynamicsInTerm(original *Expr, f *equalityFactory, term *Term, result Body) (Body, *Term) {
+func rewriteDynamicsInTerm(original *Expr, f *equalityFactory, bi map[string]*Builtin, term *Term, result Body) (Body, *Term) {
 	switch v := term.Value.(type) {
 	case Ref:
 		for i := 1; i < len(v); i++ {
-			result, v[i] = rewriteDynamicsOne(original, f, v[i], result)
+			result, v[i] = rewriteDynamicsOne(original, f, bi, v[i], result)
 		}
 	case *ArrayComprehension:
-		v.Body = rewriteDynamics(f, v.Body)
+		v.Body = rewriteDynamics(f, bi, v.Body)
 	case *SetComprehension:
-		v.Body = rewriteDynamics(f, v.Body)
+		v.Body = rewriteDynamics(f, bi, v.Body)
 	case *ObjectComprehension:
-		v.Body = rewriteDynamics(f, v.Body)
+		v.Body = rewriteDynamics(f, bi, v.Body)
 	default:
-		result, term = rewriteDynamicsOne(original, f, term, result)
+		result, term = rewriteDynamicsOne(original, f, bi, term, result)
 	}
 	return result, term
 }
 
-func rewriteDynamicsOne(original *Expr, f *equalityFactory, term *Term, result Body) (Body, *Term) {
+func rewriteDynamicsOne(original *Expr, f *equalityFactory, bi map[string]*Builtin, term *Term, result Body) (Body, *Term) {
+	return rewriteDynamicsOneKeepRefs(original, f, bi, term, result, false)
+}
+
+func rewriteDynamicsOneKeepRefs(original *Expr, f *equalityFactory, bi map[string]*Builtin, term *Term, result Body, keepRef bool) (Body, *Term) {
 	switch v := term.Value.(type) {
 	case Ref:
+		if keepRef && v.IsGround() {
```

Comment:
🔍 Compile-time rewrite: Instead of doing
```rego
__local2__ = data.foo.bar
count(__local2__)
```
we would leave this untouched:
```rego
count(data.foo.bar)
```

Reply:
So I've come to realise that doing this in the compiler isn't a good idea:

```rego
x := "baz"
count(data.foo.bar[x])
```
would be exempt here -- what we need to do is check in topdown if the function plugged ref is ground. If it is, and if the builtin is equipped to deal with Ref args, when can pass the plugged ref into the builtin.

To be able to do that, we'll need to keep all refs in function args here, and deal with the ground/non-groundness of the plugged ref in topdown.

For the IR, we'll do the same we've done here, but ad-hoc, and for all refs: introduce a local, unify, pass the local to the function.